qldpc/codes.py

```diff
@@ -210,21 +218,111 @@ def dimension(self) -> int:
+        vector: npt.NDArray[np.int_] | None = None,
+        **decoder_args: object,
+    ) -> int:
+        """Estimate the distance of this code, or the minimal weight of a nontrivial code word.
+
+        The code distance is the minimal Hamming weight between two code words, or equivalently the
+        minimal weight of a nonzero code word.
+
+        If passed a vector, compute the minimal Hamming distance between the vector and a code word.
+
+        This method uses essentially the same trick as that in CSSCode.get_one_distance_bound.
+        """
+        if bound:
+            return self.get_distance_bound(bound, vector=vector, **decoder_args)
+
+        return self.get_distance_exact(vector)
+
+    def get_distance_exact(self, vector: npt.NDArray[np.int_] | None = None) -> int:
+        """Compute the code distance by brute force.
+
+        The code distance is the minimal Hamming weight between two code words, or equivalently the
+        minimal weight of a nonzero code word.
+
+        If passed a vector, compute the minimal Hamming distance between the vector and a code word.
+        """
+        if vector is not None:
+            words = self.words() - vector[np.newaxis, :]
+        else:
+            words = self.words()[1:]
+        word_array = words.view(np.ndarray)
+        return np.min(np.count_nonzero(word_array, axis=1))
+
+    def get_distance_bound(
+        self,
+        num_trials: int,
+        vector: npt.NDArray[np.int_] | None = None,
+        **decoder_args: object,
+    ) -> int:
+        """Compute an upper bound by minimizing many individual upper bounds."""
+        return min(
+            self.get_one_distance_bound(vector=vector, **decoder_args) for _ in range(num_trials)
+        )
+
+    def get_one_distance_bound(
+        self,
+        vector: npt.NDArray[np.int_] | None = None,
+        **decoder_args: object,
+    ) -> int:
+        """Single upper bound to the distance of this code.
+
+        If passed a vector, compute an upper bound of the Hamming distance between the vector and a
+        code word.
+        """
+        _fix_decoder_args_for_nonprime_fields(decoder_args, self.field)
+
+        syndrome = self.matrix @ (vector or self.field.Zeros(self.num_bits))
+        effective_syndrome = np.append(syndrome, [1]).view(np.ndarray)
+
+        decoding_succeeded = False
+        while not decoding_succeeded:
+            word = get_random_nontrivial_vec(self.field, self.num_bits)
+            effective_check_matrix = np.vstack([self.matrix, word]).view(np.ndarray)
+
+            closest_vec = qldpc.decoder.decode(
+                effective_check_matrix,
+                effective_syndrome,
+                **decoder_args,
+            )
+
+            actual_syndrome = effective_check_matrix @ closest_vec % self.field.order
+            decoding_succeeded = np.array_equal(actual_syndrome, effective_syndrome)
+
+        return int(np.count_nonzero(closest_vec))
```

Comment:
@mittaltushant we may have discussed this before, but can you remind me why this trick works for a classical code?

Also I did a good deal of code rearrangement, so please make sure that I did not introduce a bug...

Reply:
The trick basically introduces a new constraint that prevents the vector 0 from being included in the minimum distance optimization. It may also exclude other small vectors, so it gives an upper bound. Yes, if it is None, it will work

Reply:
Cool that makes sense.  Thanks!